### ISSUE TITLE
Cannot set height to 0

### ISSUE THREAD

**Issue by domoritz** [vega/vega-lite#7798]
[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEzjQATjRyZ289AEEABBBoIcguIaZJ1h2DcyGA7nRiHETOMtXLDypJhUiioBKKigxEiCDGpoANqgYSD6wUxoAEwAHAC+ColoIABCqWhiYrn56ADCJagALADMFSBJACK1AJwAjM1JAKK1mT15LQUAYrViTSNJAOK1XR29BQASgwDsy+gAkpPp2QC6uRhwNGRYaAAMSsjqANYF1urBcLJQbMo0smRooAAevxAADMaHBBMoCiglJgAJ44OAFWRsBBfKLBJB-GgQQGCJCeQT6b4WK7ZY4wwEgsEQ9BpaFwhHoACODCQsh0gR0pBApOyQA)

Setting height to 0 is being ignored.

**Comment by horvathaa** [comment]
Hi Dominik! After our chat yesterday, I took a stab at handling this bug. Here's what I've found...

The problem seems to be in two parts of the code. The first part is in the Unit model's constructor where the `size` property is set: https://github.com/vega/vega-lite/blob/81866c0fd2a692afc6acb9c0b3a40e86074abe51/src/compile/unit.ts#L111 

Since it is just checking for `spec.height`, the 0 values are ignored. Simply switching to `spec.height !== undefined` resolves this issue. 

The second part, downstream in the compilation during parsing, is when the size property is used in `parseUnitLayoutSize` and suffers from the same issue of not checking for `undefined` as opposed to a falsey value, thus the same fix applies: https://github.com/vega/vega-lite/blob/81866c0fd2a692afc6acb9c0b3a40e86074abe51/src/compile/layoutsize/parse.ts#L100

Implementing these fixes results in the behavior I think one would expect.
![Image](https://github.com/user-attachments/assets/11dfd69d-46ed-4ad8-ae5d-8a0116296312)

However, there is a something that I wanted to ask you about. Initially, in my fix, I changed this logic for both `width` and `height`, since it seemed unintuitive to me that height would behave differently than width, thus both properties should be able to be set to 0. When using that logic, though, there was a problem where the compilation did not work, as the produced spec's scale value was wrong: `scales[0].range.step.signal: "x_step"` -- this crashed the editor. :-( 

So my question is what the design and expected behavior should be. Is it okay to ignore `width: 0`, given that the compiler does not accept that value? Or should we show an error message? Alternatively, should setting width to 0 behave similarly to `height: 0` and just make the x-axis flush with the y-axis, if it exists? But then what should we do if both height and width are 0? Just render a dot?

Let me know what you think -- I am happy to modify the code as needed and open a PR once we come to a conclusion. :-)

**Comment by domoritz** [comment]
We should try to be consistent. The hight: 0 issue may just be a specific case of a larger issue. In general, in Vega-Lite we aim to render a spec even if it has mistakes. We may drop unsupported properties or conflicting values. In many cases we raise a warning. Crashing is usually the last resort. We should also try to be consistent about width and height so figuring out the crash and fixing is would be good. 

https://github.com/vega/vega-lite/blob/81866c0fd2a692afc6acb9c0b3a40e86074abe51/src/util.ts#L534 may be useful to fix this issue.

**Comment by domoritz** [comment]
@horvathaa what do you need to send a pull request?

**Comment by domoritz** [comment]
@horvathaa checking in to see if I should hand this off

**Comment by apb-reports** [comment]
Just set height = -1
That seems to do the trick.

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v6.json",
  "description": "A simple bar chart with embedded data.",
  "height": -1,
  "data": {
    "values": [
      {"a": "A", "b": 28}, {"a": "B", "b": 55}, {"a": "C", "b": 43},
      {"a": "D", "b": 91}, {"a": "E", "b": 81}, {"a": "F", "b": 53},
      {"a": "G", "b": 19}, {"a": "H", "b": 87}, {"a": "I", "b": 52}
    ]
  },
  "mark": "bar",
  "encoding": {
    "x": {"field": "a", "type": "nominal", "axis": {"labelAngle": 0}},
    "y": {"field": "b", "type": "quantitative"}
  }
}
```

**Comment by domoritz** [comment]
Yep, that's a workaround but not a fix. I don't think many people want to have a chart of height 0 anyway but it's good to fix this.

**Comment by horvathaa** [comment]
Hi @domoritz -- sorry for the lack of response, I missed your initial @ a few weeks ago. I'm happy to finish my PR, but I'm a bit slammed for the next two weeks or so with the UIST deadline + personal travel. So, if you think this is pressing enough to hand off, I am okay with that. If you are okay waiting, I just want to clarify the next steps, assuming that I understood your response properly, including intended behavior:

- Ensure that `hasProperty` is not the source of this "treating 0 as false" bug.
- Make `width` and `height` behave consistently, thus both can be set to 0.
- Perhaps show a warning that setting width or height to 0 is an anti-pattern/mistake?
